### ISSUE TITLE
unable to explain cnn text clf model  using lime

### ISSUE THREAD

**Issue by swayam01** [marcotcr/lime#260]
Thanks for the giving us lime.
I am currently using word embedding + cnn for multi-class text classification
But, getting ValueError: Found input variables with inconsistent numbers of samples: [24, 1]
When I run :
explainer.explain_instance(text_instance=‘To add ?relieve in productive cough? statements in Indication section’, labels=[0,1,2,3,4],classifier_fn=new_predict, num_features=6, num_samples=24)

My model architecture:

I have wrapped the predict function into a function that takes in a list of strings and returns prediction probabilities.
def new_predict(texts):
ipt = texts
SoC = StringIO(ipt[0])
df = pd.read_csv(SoC,names=[‘Text’])
df.Text = df.Text.apply(lambda x: x.lower())
mystring = re.sub(r"^\W+", “”, df.Text[0])
mystring = StringIO(mystring)
train_df = pd.read_csv(mystring,names=[‘Text’])
undesirable_ascii_characters = list(range(32))
undesirable_ascii_characters.remove(10) #keep new line since this might be used for sentence tokenizer
undesirable_charmap = dict.fromkeys(undesirable_ascii_characters)
train_df.Text = train_df.Text.apply(lambda s: s.lower().encode(‘ascii’, ‘ignore’).decode(‘utf-8’, ‘ignore’))
train_df.Text = train_df.Text.apply(lambda s: str(s).translate(undesirable_charmap))
train_df.Text = train_df.Text.apply(lambda s: s.replace(u’<u+201c>’, ‘’))
train_df.Text = train_df.Text.apply(lambda s: s.replace(u’<u+201d>’, ‘’))
train_df.Text = train_df.Text.apply(lambda s: s.replace(u’<u+2022>’, ‘’))
train_df.Text = train_df.Text.apply(lambda s: s.replace(u’<u+2019>’, ‘’))
train_df.Text = train_df.Text.apply(lambda s: s.replace(u’<u+2013>’, ‘’))
train_df.Text = train_df.Text.apply(lambda s: s.replace(u’<u+2018>’, ‘’))
train_df.Text = train_df.Text.apply(lambda s: s.replace(u’<u+00b0>’, ‘’))
train_df.Text = train_df.Text.apply(lambda s: s.replace(u’> ‘, ‘’))
train_df.Text = train_df.Text.apply(lambda s: s.replace(u’& ‘, ‘’))
train_df.Text = train_df.Text.apply(lambda s: s.replace(u’ ‘, ‘’))
train_df.Text = train_df.Text.apply(lambda s: apply_custom_regx(s))
train_df.Text = train_df.Text.apply(lambda s: s.replace(’\n’, ‘’))
train_df[‘Sentences’] = train_df.Text.apply(lambda s: sent_tokenize(s))
train_df.Sentences = train_df.Sentences.apply(lambda sentences: [word_tokenize(sentence) for sentence in sentences])
clean_start_end = lambda doc: [[re.sub(’[a-zA-z0-9]|[^a-zA-Z0-9]$’,’’,w) for w in sent] for sent in doc]
train_df.Sentences = train_df.Sentences.apply(clean_start_end)
train_df.Sentences = train_df.Sentences.apply(lambda doc: [[w for w in sent if w] for sent in doc]) #Trail an Test
train_df.drop([“Text”], axis=1, inplace=True)
gen_data = pred_global_utils.GenerateDataset(train_df, corpus_wordidx)
x_train_22_T = gen_data.generate_data(custom_unit_dict, has_class=True, add_start_end_tag=True)
del gen_data
from pred_global_utils import oov_tk
word_unknown_tag_idx = corpus_wordidx["<UNK>"]
char_unknown_tag_idx = pred_global_utils.char_unknown_tag_idx
MAX_TEXT_LEN = 24
x_train_22_T = pad_sequences(x_train_22_T, maxlen=MAX_TEXT_LEN, value=word_unknown_tag_idx,
padding=“post”,truncating=“post”)
return model_1.predict(x_train_22_T)

wheni do like: new_predict([‘To add ?relieve in productive cough? statements in Indication section’])
I get the expected result but when I do: exp = explainer.explain_instance(‘To add ?relieve in productive cough? statements in Indication section’, new_predict, num_features=6,num_samples=24,labels=[0, 4])

Not sure what is happening wrong here ? a bit confusing about how to rewrite the wrapper to make it eligible for the signature of explain_instance function.
Thank you in advance for your advice.

**Comment by swayam01** [comment]
Got Rid of the value error using: 
exp = explainer.explain_instance(txt, new_predict, num_features=6, num_samples=1, labels=[0,1,2,3,4])

But still didn't got the result. my output now using as exp.as_list() is:
[('HA', 0.0),
 ('requested', 0.0),
 ('label', 0.0),
 ('update', 0.0),
 ('following', 0.0),
 ('decision', 0.0)]

dont understand why all weight are 0.0. what i am doing wrong here. 
thanks in advance

**Comment by ian-perry-ironnet** [comment]
I have a slightly different situation but same result, all `0.0`'s for the explanation effects.  I'm using `LimeTabularExplainer`.  What could be causing this to happen?

**Comment by marcotcr** [comment]
you need `num_samples` to be a reasonable amount (let's say `100`, `1000`, `5000`...). In this case, you are getting an explanation from **one** perturbed sample, so you basically just get the intercept as an explanation.
My guess is that your original error is in your wrapped predict function, which is unreadable in your comment above.

**Comment by swayam01** [comment]
Yes that's where i am struggling. please find my code below with the error:
`txt = 'requested label update following decision in all hydrochlorothiazide containing products regarding the risk of non-melanoma skin cancer'`
`txt = np.array([txt], dtype='str')`
`Text = df['Text']
df.label = df.label.apply(lambda x: x.lower())
labels = df['label']
np.random.seed(0)
idx = np.random.permutation(len(Text))
Text,labels = np.asarray(Text)[idx],np.asarray(labels)[idx]
texts = np.asarray(Text, dtype='U1055')`
`class_names = list(np.unique(labels))
explainer = LimeTextExplainer(class_names=class_names,feature_selection="auto")`
`def new_predict(texts):
    sentences = sent_tokenize(texts)
    for sentence in texts:
        doc = [word_tokenize(sentence)]
    for sent in doc:
        s = [[w for w in sent if w]]
    sent2id = []
    for word in s[0]:
        try:
            if(corpus_wordidx[word] < vocab_size):
                sent2id = sent2id + [corpus_wordidx[word]]
            else:
                sent2id = sent2id + [word_unknown_tag_idx]
        except KeyError as e:
            pass
    x_train_22_T = pad_sequences([sent2id], maxlen=MAX_TEXT_LEN, value=word_unknown_tag_idx,padding="post",truncating="post")
    return model_1.predict(x_train_22_T)`
`new_predict(Text[0])`

I get the list of prob using new_predict(Text[0])
but when i do `exp = explainer.explain_instance(txt, new_predict, num_features=6, num_samples=1, labels=[0,1,2,3,4])`
i get error: `TypeError: cannot use a string pattern on a bytes-like object`

Also if i fix this i get error `ValueError: Found input variables with inconsistent numbers of samples: [100, 1]` for when i change num_samples = 100.

Please review and support me. Thanks

**Comment by swayam01** [comment]
I am able to resolve the error using this wrapper function:
`def new_predict(texts):`
    `_seq = tokenizer.texts_to_sequences(texts)`
    `_text_data = pad_sequences(_seq, maxlen=24)`
    `return cnn.predict(_text_data)`

i want to understand why it does not worked with custom wrapper function and gave results using keras api

**Comment by swayam01** [comment]
please take a look at the wrapped predict function
`def new_predict(texts):`
    `sentences = [word_tokenize(i) for i in txt]`
    `sent2id = []`
    `for word in sentences[0]:`
        `try:`
            `if(corpus_wordidx[word] < vocab_size):`
                `sent2id = sent2id + [sentence_start_tag_idx]`
                `sent2id = sent2id + [corpus_wordidx[word]]`
                `sent2id = sent2id + [sentence_end_tag_idx]`
            `else:`
               `sent2id = sent2id + [word_unknown_tag_idx]`
                `sent2id = sent2id + [sentence_end_tag_idx]`
        `except KeyError as e:`
            `pass`
        `x_train_22_T = pad_sequences([sent2id], maxlen=MAX_TEXT_LEN, value=word_unknown_tag_idx,padding="post",truncating="post")`
        `return model_1.predict(x_train_22_T)`
 here i have my texts as list of strings. I get error with `num_samples` 100
`ValueError: Found input variables with inconsistent numbers of samples: [100, 1]`

please help me debug.
Thank-You
